File: jobs/java_transform/src/main/java/Transform.java
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.expressions.UserDefinedFunction;
import org.apache.spark.sql.types.DataTypes;
import scala.Tuple2;

import java.io.*;
import java.util.Map;

import static org.apache.spark.sql.functions.*;

public class Transform implements Serializable {
	String filePath;

	SparkSession ss;

	Dataset<Row> items;

	Map<String, Double> currency;
	Map<Long, Area> areas;

	public Transform() {
		this.filePath = System.getenv("filePath");
		init();
	}

	public void init() {
		ss = SparkSession.builder().appName("Transform Job").master("local[*]").getOrCreate();

		items = ss.read().json(filePath + System.getenv("date") + "vacancies*");


		// currency
		currency = ss.read().json(filePath + "dictionaries")
				.withColumn("currency", explode(col("currency")))
				.select("currency.*")
				.toJavaRDD()
				.mapToPair((PairFunction<Row, String, Double>) row -> {
					String key = row.getString(row.fieldIndex("code"));
					Double rate = row.getDouble(row.fieldIndex("rate"));
					return new Tuple2<>(key, rate);
				}).collectAsMap();

		/*JavaRDD<Row> curr = ss.read().json(filePath + "dictionaries")
				.withColumn("currency", explode(col("currency")))
				.select("currency.*")
				.toJavaRDD();
		JavaPairRDD<String, Double> curMap = curr.mapToPair((PairFunction<Row, String, Double>) row -> {
            		String key = row.getString(row.fieldIndex("code"));
            		Double rate = row.getDouble(row.fieldIndex("rate"));
            		return new Tuple2<>(key, rate);
        });
		currency = curMap.collectAsMap();*/


		// areas
		areas = ss.read().json(filePath + "areas").toJavaRDD()
				.mapToPair((PairFunction<Row, Long, Area>) row -> {
					Long id = row.getLong(row.fieldIndex("id"));
					Area area = new Area();
					area.name = row.getString(row.fieldIndex("name"));
					if (row.isNullAt(row.fieldIndex("parent_id")))
						area.parent_id = null;
					else
						area.parent_id = row.getLong(row.fieldIndex("parent_id"));
					return new Tuple2<>(id, area);
				}).collectAsMap();

		/*JavaRDD<Row> ars = ss.read().json(filePath + "areas").toJavaRDD();

		JavaPairRDD<Long, Area> arsMap = ars.mapToPair((PairFunction<Row, Long, Area>) row -> {
			Long id = row.getLong(row.fieldIndex("id"));
			Area area = new Area();
			area.name = row.getString(row.fieldIndex("name"));
			if (row.isNullAt(row.fieldIndex("parent_id")))
				area.parent_id = null;
			else
				area.parent_id = row.getLong(row.fieldIndex("parent_id"));
			return new Tuple2<>(id, area);
		});
		areas = arsMap.collectAsMap();*/
	}

	public void transform() {

		Dataset<Row> finals = items
				.withColumn("id", col("id").cast(DataTypes.LongType))

				.withColumn("region_area_id", getString.apply(col("area"),
						lit("id")).cast(DataTypes.LongType))
				.withColumn("country_area_id", country.apply(col("area")))

				.withColumn("currency", getString.apply(col("salary"),
						lit("currency")))
				.withColumn("salary_from_local", getLong.apply(col("salary"),
						lit("from")))
				.withColumn("salary_to_local", getLong.apply(col("salary"),
						lit("to")))
				.withColumn("salary_from_rub", salary_rub.apply(col("salary"),
						lit("from")))
				.withColumn("salary_to_rub",  salary_rub.apply(col("salary"),
						lit("to")))

				.withColumn("close_to_metro", close_to_metro.apply(col("address")))

				/*.withColumn("requirement", getString.apply(col("snippet"),
						lit("requirement")))
				.withColumn("responsibility", getString.apply(col("snippet"),
						lit("responsibility")))*/

				.withColumn("schedule_id", getString.apply(col("schedule"),
						lit("id")))
				.withColumn("experience_id", getString.apply(col("experience"),
						lit("id")))
				.withColumn("employment_id", getString.apply(col("employment"),
						lit("id")))

				.select("id", "region_area_id", "country_area_id", "currency",
						"salary_from_local", "salary_to_local", "salary_from_rub", "salary_to_rub",
						"close_to_metro", "schedule_id", "experience_id", "employment_id")

				.dropDuplicates("id");

		finals.write()
				.mode("overwrite")
				.parquet(filePath + System.getenv("date") + "transformed");

        ss.stop();
	}

	// default
	UserDefinedFunction getLong = udf((Row row, String field) -> {
		if (row == null) return null;
		int i = row.fieldIndex(field);

		if (row.isNullAt(i)) return null;
		return row.getLong(i);
	}, DataTypes.LongType);

	UserDefinedFunction getString = udf((Row row, String field) -> {
		if (row == null) return null;
		int i = row.fieldIndex(field);

		if (row.isNullAt(i)) return null;
		return row.getString(i);
	}, DataTypes.StringType);

	// salary in rub
	UserDefinedFunction salary_rub = udf((Row salary, String field) -> {
		if (salary == null) return null;
		int c = salary.fieldIndex("currency");
		int i = salary.fieldIndex(field);

		if (salary.isNullAt(c) || salary.isNullAt(i)) return null;

		double rate = currency.get(salary.getString(c));
        return (long) (salary.getLong(i) / rate);
	}, DataTypes.LongType);

	// metro
	UserDefinedFunction close_to_metro = udf((Row address) -> {
		if (address == null) return null;
		int i = address.fieldIndex("metro");
		return !address.isNullAt(i);
 	}, DataTypes.BooleanType);

	// country
	UserDefinedFunction country = udf((Row area) -> {
		if (area == null) return null;

		long id = Long.parseLong(area.getString(area.fieldIndex("id")));
		Area curr = areas.get(id);

		while (curr.parent_id != null) {
			id = curr.parent_id;
			curr = areas.get(curr.parent_id);
		}

		return id;
	}, DataTypes.LongType);
}

File: jobs/scala_transform/src/main/scala/Transform.scala
import org.apache.spark.sql.expressions.UserDefinedFunction
import org.apache.spark.sql.functions.{col, explode, udf, lit}
import org.apache.spark.sql.{Dataset, Row, SparkSession}

case class Area(name: String, parent_id: Option[Long])

// val area = Area.upply

class Transform(filePath: String, ss: SparkSession) extends Serializable {
  private val items: Dataset[Row] = ss.read.json(s"$filePath${sys.env("date")}vacancies*")

  private var currency: Map[String, Double] = _
  private var areas: Map[Long, Area] = _

  def this() = {
    this(
      sys.env("filePath"),
      SparkSession
        .builder()
        .appName("Transform Job")
        .master("local[*]")
        .getOrCreate())

    // currency
    currency = ss.read.
      json(s"${filePath}dictionaries").
      withColumn("currency", explode(col("currency"))).
      select("currency.*").rdd.
      map(row => {
        val key = row.getAs[String]("code")
        val rate = row.getAs[Double]("rate")
        key -> rate
      }).collectAsMap().toMap

    // areas
    areas = ss.read.
      json(s"${filePath}areas").rdd.
      map(row => {
        val id = row.getAs[Long]("id")
        val name = row.getAs[String]("name")
        val parent_id = Option(row.getAs[Long]("parent_id"))
        id -> Area(name, parent_id)
      }).collectAsMap().toMap
  }

  def main(args: Array[String]): Unit = {
    val transform = new Transform()
    transform.transform()
  }

  private def transform(): Unit = {
    val finals = items
      .withColumn("id", col("id").cast("long"))
      .withColumn("region_area_id", getStringUDF(col("area"), lit("id")).cast("long"))
      .withColumn("country_area_id", countryUDF(col("area")))
      .withColumn("currency", getStringUDF(col("salary"), lit("currency")))
      .withColumn("salary_from_local", getLongUDF(col("salary"), lit("from")))
      .withColumn("salary_to_local", getLongUDF(col("salary"), lit("to")))
      .withColumn("salary_from_rub", salaryRubUDF(col("salary"), lit("from")))
      .withColumn("salary_to_rub", salaryRubUDF(col("salary"), lit("to")))
      .withColumn("close_to_metro", closeToMetroUDF(col("address")))
      .withColumn("schedule_id", getStringUDF(col("schedule"), lit("id")))
      .withColumn("experience_id", getStringUDF(col("experience"), lit("id")))
      .withColumn("employment_id", getStringUDF(col("employment"), lit("id")))
      .select(
        "id", "region_area_id", "country_area_id", "currency",
        "salary_from_local", "salary_to_local", "salary_from_rub", "salary_to_rub",
        "close_to_metro", "schedule_id", "experience_id", "employment_id"
      )
      .dropDuplicates("id")

    finals.write
      .mode("overwrite")
      .parquet(s"$filePath${sys.env("date")}transformed")

    ss.stop()
  }

  private val getLongUDF: UserDefinedFunction = udf((row: Row, field: String) => {
    Option(row).flatMap(r => {
      val idx = r.fieldIndex(field)
      if (!r.isNullAt(idx)) Option(r.getAs[Long](idx)) else None
    })
  })

  private val getStringUDF: UserDefinedFunction = udf((row: Row, field: String) => {
    Option(row).flatMap(r => {
      val idx = r.fieldIndex(field)
      if (!r.isNullAt(idx)) Option(r.getAs[String](idx)) else None
    })
  })

  private val closeToMetroUDF: UserDefinedFunction = udf((address: Row) => {
    Option(address).exists(a => {
      val metroIdx = a.fieldIndex("metro")
      !a.isNullAt(metroIdx)
    })
  })

  private val salaryRubUDF: UserDefinedFunction = udf((salary: Row, field: String) => {
    Option(salary).flatMap(s => {
      val currencyIdx = s.fieldIndex("currency")
      val fieldIdx = s.fieldIndex(field)
      if (!s.isNullAt(currencyIdx) && !s.isNullAt(fieldIdx)) {
        val rate = currency.getOrElse(s.getAs[String](currencyIdx), 1.0)
        Option((s.getAs[Long](fieldIdx) / rate).toLong)
      } else None
    })
  })

  private val countryUDF: UserDefinedFunction = udf((area: Row) => {
    Option(area).flatMap(a => {
      var id = a.getAs[String]("id").toLong
      var currentArea = areas.get(id)
      while (currentArea.exists(_.parent_id.isDefined)) {
        id = currentArea.flatMap(_.parent_id).get
        currentArea = areas.get(id)
      }
      Option(id)
    })
  })
}